File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] valeurs) {
        int debutAchat = 0;
        int profitMaximal = 0;
        
        for (int i = 1; i < valeurs.length; i++) {
            if (valeurs[debutAchat] < valeurs[i]) {
                profitMaximal = Math.max(profitMaximal, valeurs[i] - valeurs[debutAchat]);
            } else {
                debutAchat = i;
            }
        }
        
        return profitMaximal;
    }
}

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.py
class Solution(object):
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        profit =0
        l,r=0,1
    
        while r < len(prices):
            
            if prices[r]-prices[l]>profit:
                profit=prices[r]-prices[l]
            
            if prices[r]<prices[l]:
                l=r
            r+=1
        return profit